packages/flutter/lib/src/widgets/heroes.dart

```diff
@@ -166,299 +134,270 @@ class Hero extends StatefulWidget {
-  }
-}
+  final _OnFlightEnded onFlightEnded;
 
-class _HeroMatch {
-  const _HeroMatch(this.from, this.to, this.tag);
-  final _HeroHandle from;
-  final _HeroHandle to;
-  final Object tag;
-}
+  RectTween heroRect;
+  Animation<double> _heroOpacity = kAlwaysCompleteAnimation;
+  ProxyAnimation _proxyAnimation;
+  _HeroFlightManifest manifest;
+  OverlayEntry overlayEntry;
 
-/// Signature for a function that takes two [Rect] instances and returns a
-/// [RectTween] that transitions between them.
-///
-/// This is typically used with a [HeroController] to provide an animation for
-/// [Hero] positions that looks nicer than a linear movement. For example, see
-/// [MaterialRectArcTween].
-typedef RectTween CreateRectTween(Rect begin, Rect end);
+  RectTween _doCreateRectTween(Rect begin, Rect end) {
+    if (manifest.createRectTween != null)
+      return manifest.createRectTween(begin, end);
+    return new RectTween(begin: begin, end: end);
+  }
 
-class _HeroParty {
-  _HeroParty({ this.onQuestFinished, this.createRectTween });
+  // The OverlayEntry WidgetBuilder callback for the hero's overlay.
+  Widget _buildOverlay(BuildContext context) {
+    assert(manifest != null);
+    return new AnimatedBuilder(
+      animation: _proxyAnimation,
+      child: manifest.toHero.config,
+      builder: (BuildContext context, Widget child) {
+        final RenderBox toHeroBox = manifest.toHero.context?.findRenderObject();
+        if (toHeroBox == null || !toHeroBox.attached) {
+          // The toHero no longer exists. Continue flying while fading out.
+          if (_heroOpacity == kAlwaysCompleteAnimation) {
+            _heroOpacity = new Tween<double>(begin: 1.0, end: 0.0)
+              .chain(new CurveTween(curve: new Interval(_proxyAnimation.value, 1.0)))
+              .animate(_proxyAnimation);
+          }
+        } else if (toHeroBox.hasSize) {
+          // The toHero has been laid out. If it's no longer where the hero animation is
+          // supposed to end up (heroRect.end) then recreate the heroRect tween.
+          final RenderBox routeBox = manifest.toRoute.subtreeContext?.findRenderObject();
+          final Point heroOriginEnd = toHeroBox.localToGlobal(Point.origin, ancestor: routeBox);
+          if (heroOriginEnd != heroRect.end.topLeft) {
+            final Rect heroRectEnd = heroOriginEnd & heroRect.end.size;
+            heroRect = _doCreateRectTween(heroRect.begin, heroRectEnd);
+          }
+        }
+
+        final Rect rect = heroRect.evaluate(_proxyAnimation);
+        final Size size = manifest.navigatorRect.size;
+        final RelativeRect offsets = new RelativeRect.fromSize(rect, size);
+
+        return new Positioned(
+          top: offsets.top,
+          right: offsets.right,
+          bottom: offsets.bottom,
+          left: offsets.left,
+          child:  new IgnorePointer(
+            child: new RepaintBoundary(
+              child: new Opacity(
+                key: manifest.toHero._key,
```

Comment:
Hey, I'm doing the refactor we were talking about. I didn't really get what this is for though. We're taking the Key from the Hero widget instance and putting it as a key for this Opacity widget. What is it meant to do?

Reply:
Making sure that the same actual widget subtree is used, rather than recreating widgets each time. That way anything stateful going on is maintained across the transition.